File: python/demos/demo_glb_viewer.py
"""Demo: GLB Model Viewer (Python)

Python mirror of demos/demo_glb_viewer.cpp.

Interactive viewer for glTF 2.0 binary (.glb) 3D models.
Loads all .glb files from data/glb_files/ and lets you switch between them.

Controls:
  LEFT/RIGHT ARROW - Switch between models
  1-8              - Select model directly
  B                - Toggle bounding box display
  +/-              - Adjust model transparency

Note: wireframe mode is not available through the Python API.

Run from the repo root:
    python python/demos/demo_glb_viewer.py
"""

import glob
import os
import geodraw
from pathlib import Path


def find_glb_files(dir_path):
    """Return sorted list of .glb file paths in dir_path."""
    pattern = os.path.join(dir_path, "*.glb")
    return sorted(glob.glob(pattern))


def main():
    app = geodraw.App(1280, 720, "Demo GLB Viewer")

    root = os.environ.get("GEODRAW_ROOT")
    if not root:
        raise RuntimeError("GEODRAW_ROOT not set")

    glb_dir = Path(root) / "data" / "glb_files"
    paths = find_glb_files(glb_dir)

    if not paths:
        print(f"Error: No .glb files found in {glb_dir}")
        print("Please ensure .glb files exist in data/glb_files/ directory")
        return

    print(f"Found {len(paths)} .glb files")

    # Load all models upfront
    models = []
    for path in paths:
        model = app.load_glb(path)
        if model is not None and model.is_valid:
            models.append(model)
            print(f"Loaded: {model.name}")
        else:
            print(f"Failed to load: {path}")
            models.append(None)

    valid_models = [m for m in models if m is not None]
    if not valid_models:
        print("Error: No models could be loaded")
        return

    # Mutable state (use lists so lambdas can mutate)
    state = {
        "index": 0,
        "alpha": [1.0],
        "show_bounds": [True],
    }

    def current_model():
        return models[state["index"]]

    def next_model():
        state["index"] = (state["index"] + 1) % len(models)
        m = current_model()
        print(f"Switched to: {m.name if m else '(invalid)'}")

    def prev_model():
        state["index"] = (state["index"] - 1) % len(models)
        m = current_model()
        print(f"Switched to: {m.name if m else '(invalid)'}")

    def select_model(idx_1based):
        idx = idx_1based - 1
        if 0 <= idx < len(models):
            state["index"] = idx
            m = current_model()
            print(f"Selected model {idx_1based}: {m.name if m else '(invalid)'}")

    def increase_alpha():
        state["alpha"][0] = min(1.0, state["alpha"][0] + 0.1)
        print(f"Model alpha: {state['alpha'][0]:.1f}")

    def decrease_alpha():
        state["alpha"][0] = max(0.0, state["alpha"][0] - 0.1)
        print(f"Model alpha: {state['alpha'][0]:.1f}")

    def toggle_bounds():
        state["show_bounds"][0] = not state["show_bounds"][0]

    # Register commands
    app.add_cmd("next-model",  next_model,  "Switch to next model",     key=geodraw.Key.RIGHT)
    app.add_cmd("prev-model",  prev_model,  "Switch to previous model", key=geodraw.Key.LEFT)

    app.add_cmd("select-model-1", lambda: select_model(1), "Select model 1", key=geodraw.Key.D1)
    app.add_cmd("select-model-2", lambda: select_model(2), "Select model 2", key=geodraw.Key.D2)
    app.add_cmd("select-model-3", lambda: select_model(3), "Select model 3", key=geodraw.Key.D3)
    app.add_cmd("select-model-4", lambda: select_model(4), "Select model 4", key=geodraw.Key.D4)
    app.add_cmd("select-model-5", lambda: select_model(5), "Select model 5", key=geodraw.Key.D5)
    app.add_cmd("select-model-6", lambda: select_model(6), "Select model 6", key=geodraw.Key.D6)
    app.add_cmd("select-model-7", lambda: select_model(7), "Select model 7", key=geodraw.Key.D7)
    app.add_cmd("select-model-8", lambda: select_model(8), "Select model 8", key=geodraw.Key.D8)

    app.add_cmd("increase-alpha", increase_alpha, "Increase transparency", key=geodraw.Key.EQUAL)
    app.add_cmd("decrease-alpha", decrease_alpha, "Decrease transparency", key=geodraw.Key.MINUS)
    app.add_cmd("toggle-bounds",  toggle_bounds,  "Toggle bounding box",   key=geodraw.Key.B)

    # Frame camera to first valid model
    first = current_model()
    if first is not None:
        bounds = first.bounds
        app.auto_frame(bounds[0], bounds[1])

    app.request_continuous_update("glb_viewer.display")

    @app.add_update_callback()
    def update(dt):
        scene = app.scene()
        scene.clear()

        model = current_model()
        if model is None or not model.is_valid:
            return

        scene.add_loaded_model(model, alpha=state["alpha"][0])

        if state["show_bounds"][0]:
            bounds = model.bounds
            scene.add_bbox(bounds[0], bounds[1],
                           color=(0.0, 1.0, 0.0), thickness=2.0)

    app.run()


if __name__ == "__main__":
    main()

File: demos/demo_glb_viewer.cpp
/*******************************************************************************
 * Demo: GLB Model Viewer
 *
 * Interactive viewer for glTF 2.0 binary (.glb) 3D models.
 * Loads all .glb files from data/glb_files/ and lets you switch between them.
 *
 * Controls:
 *   LEFT/RIGHT ARROW - Switch between models
 *   1-8              - Select model directly
 *   W                - Toggle wireframe mode
 *   B                - Toggle bounding box display
 *   +/-              - Adjust model transparency
 *
 * Run from the build directory:
 *   ./bin/demo_glb_viewer
 *******************************************************************************/

#include "geodraw/app/app.hpp"
#include "geodraw/scene/scene.hpp"
#include "geodraw/geometry/colors.hpp"
#include "geodraw/modules/gltf/gltf_loader.hpp"
#include <iostream>
#include <filesystem>
#include <algorithm>

using namespace geodraw;

/**
 * GlbViewerState - Application state for the GLB model viewer
 *
 * Manages collection of loaded models, display settings, and the scene queue.
 */
struct GlbViewerState {
    std::vector<gltf::LoadedModel> models;
    std::vector<std::string> modelPaths;
    int currentModelIndex = 0;

    // Display options
    bool wireframeMode = false;
    bool showBounds = true;
    float modelAlpha = 1.0f;

    // Scene queue owned here (not static)
    Scene mainQueue;

    bool hasModels() const { return !models.empty(); }
    gltf::LoadedModel& currentModel() { return models[currentModelIndex]; }
    const gltf::LoadedModel& currentModel() const { return models[currentModelIndex]; }
};

std::vector<std::string> findGlbFiles(const std::string& dirPath) {
    std::vector<std::string> glbFiles;

    try {
        if (!std::filesystem::exists(dirPath)) {
            std::cerr << "Directory not found: " << dirPath << "\n";
            return glbFiles;
        }

        for (const auto& entry : std::filesystem::directory_iterator(dirPath)) {
            if (entry.is_regular_file()) {
                std::string path = entry.path().string();
                if (path.size() >= 4 && path.substr(path.size() - 4) == ".glb") {
                    glbFiles.push_back(path);
                }
            }
        }

        std::sort(glbFiles.begin(), glbFiles.end());

    } catch (const std::exception& e) {
        std::cerr << "Error scanning directory: " << e.what() << "\n";
    }

    return glbFiles;
}

void loadAllModels(GlbViewerState& state, const std::string& dirPath, Renderer& renderer) {
    state.modelPaths = findGlbFiles(dirPath);

    if (state.modelPaths.empty()) {
        std::cerr << "No .glb files found in: " << dirPath << "\n";
        return;
    }

    std::cout << "Found " << state.modelPaths.size() << " .glb files\n";
    std::cout << "Loading models...\n";

    for (const auto& path : state.modelPaths) {
        auto model = gltf::loadGLB(path, renderer);
        if (model) {
            state.models.push_back(std::move(*model));
        } else {
            std::cerr << "Failed to load: " << path << "\n";
            state.models.push_back(gltf::LoadedModel{});
        }
    }

    std::cout << "Successfully loaded " << state.models.size() << " models\n";
}

void cleanupAllModels(GlbViewerState& state, Renderer& renderer) {
    for (auto& model : state.models) {
        if (model.isValid()) {
            gltf::unloadModel(model, renderer);
        }
    }
    state.models.clear();
    state.modelPaths.clear();
}

void updateGeometry(const GlbViewerState& state, Scene& queue) {
    queue.clear();

    if (!state.hasModels()) return;

    const auto& model = state.currentModel();
    if (!model.isValid()) return;

    for (size_t i = 0; i < model.meshes.size(); ++i) {
        GLuint tex = i < model.textureHandles.size() ? model.textureHandles[i] : 0;
        glm::vec3 color = i < model.colors.size() ? model.colors[i] : glm::vec3(1.0f);

        queue.Add(model.meshes[i],
                 /*thickness=*/1.0f,
                 LineStyle::Solid,
                 /*alpha=*/state.modelAlpha,
                 /*color=*/color,
                 /*depthTest=*/true,
                 /*filled=*/!state.wireframeMode,
                 /*textureHandle=*/tex);
    }

    if (state.showBounds) {
        queue.Add(model.bounds,
                  /*thickness=*/2.0f,
                 LineStyle::Solid,
                 /*alpha=*/1.0f,
                 /*color=*/Colors::GREEN,
                 /*depthTest=*/true, /*filled=*/false);
    }
}

void nextModel(GlbViewerState& state) {
    if (!state.hasModels()) return;
    state.currentModelIndex = (state.currentModelIndex + 1) % state.models.size();
    std::cout << "Switched to model: " << state.currentModel().name << "\n";
}

void prevModel(GlbViewerState& state) {
    if (!state.hasModels()) return;
    state.currentModelIndex--;
    if (state.currentModelIndex < 0) {
        state.currentModelIndex = state.models.size() - 1;
    }
    std::cout << "Switched to model: " << state.currentModel().name << "\n";
}

void selectModel(GlbViewerState& state, int index) {
    if (!state.hasModels()) return;
    int idx = index - 1;
    if (idx >= 0 && idx < static_cast<int>(state.models.size())) {
        state.currentModelIndex = idx;
        std::cout << "Selected model " << index << ": " << state.currentModel().name << "\n";
    }
}

void adjustAlpha(GlbViewerState& state, float delta) {
    state.modelAlpha = std::clamp(state.modelAlpha + delta, 0.0f, 1.0f);
    std::cout << "Model alpha: " << state.modelAlpha << "\n";
}

int main() {
    App app(1280, 720, "Demo GLB Viewer");

    GlbViewerState state;

    const char *root = getenv("GEODRAW_ROOT");
    if(!root){
      throw std::runtime_error("GEODRAW_ROOT not set");
    }
    std::string glb_path = std::string(root) + "/data/glb_files";

    loadAllModels(state, glb_path, app.getRenderer());

    if (!state.hasModels()) {
      std::cerr << "Error: No .glb models could be loaded from " << glb_path << std::endl;
        std::cerr << "Please ensure .glb files exist in " << glb_path << std::endl;
        return 1;
    }

    app.addCmd("next-model", [&](){ nextModel(state); },
               "Switch to next model", GLFW_KEY_RIGHT);
    app.addCmd("prev-model", [&](){ prevModel(state); },
               "Switch to previous model", GLFW_KEY_LEFT);

    app.addCmd("select-model-1", [&](){ selectModel(state, 1); }, "Select model 1", GLFW_KEY_1);
    app.addCmd("select-model-2", [&](){ selectModel(state, 2); }, "Select model 2", GLFW_KEY_2);
    app.addCmd("select-model-3", [&](){ selectModel(state, 3); }, "Select model 3", GLFW_KEY_3);
    app.addCmd("select-model-4", [&](){ selectModel(state, 4); }, "Select model 4", GLFW_KEY_4);
    app.addCmd("select-model-5", [&](){ selectModel(state, 5); }, "Select model 5", GLFW_KEY_5);
    app.addCmd("select-model-6", [&](){ selectModel(state, 6); }, "Select model 6", GLFW_KEY_6);
    app.addCmd("select-model-7", [&](){ selectModel(state, 7); }, "Select model 7", GLFW_KEY_7);
    app.addCmd("select-model-8", [&](){ selectModel(state, 8); }, "Select model 8", GLFW_KEY_8);

    app.addToggle("", state.wireframeMode, "Toggle wireframe mode", GLFW_KEY_W);
    app.addToggle("", state.showBounds,    "Toggle bounding box",   GLFW_KEY_B);

    app.addCmd("increase-alpha", [&](){ adjustAlpha(state,  0.1f); }, "Increase transparency", GLFW_KEY_EQUAL);
    app.addCmd("decrease-alpha", [&](){ adjustAlpha(state, -0.1f); }, "Decrease transparency", GLFW_KEY_MINUS);

    // Frame camera to show the first model
    {
        const auto& bounds = state.currentModel().bounds;
        glm::vec3 minPt(bounds.min.pos.x, bounds.min.pos.y, bounds.min.pos.z);
        glm::vec3 maxPt(bounds.max.pos.x, bounds.max.pos.y, bounds.max.pos.z);
        app.camera.autoFrame(minPt, maxPt, glm::radians(-60.0f), glm::radians(30.0f));
    }

    app.requestContinuousUpdate("glb_viewer.display");

    app.addUpdateCallback([&](float /*dt*/) {
        updateGeometry(state, state.mainQueue);
    });

    app.addDrawCallback([&]() {
        app.getRenderer().render(state.mainQueue, app.camera);

        if (state.hasModels()) {
            const auto& model = state.currentModel();
            std::string info = "Model: " + model.name + " (" +
                             std::to_string(state.currentModelIndex + 1) + "/" +
                             std::to_string(state.models.size()) + ")";
            app.getRenderer().renderText(info, 10, 690, glm::vec3(1.0f, 1.0f, 1.0f));

            std::string meshInfo = "Meshes: " + std::to_string(model.meshes.size()) +
                                 "  Textures: " + std::to_string(model.textureHandles.size());
            app.getRenderer().renderText(meshInfo, 10, 670, glm::vec3(0.8f, 0.8f, 0.8f));

            if (state.wireframeMode) {
                app.getRenderer().renderText("Wireframe Mode", 10, 650, glm::vec3(1.0f, 0.5f, 0.0f));
            }
        }
    });

    app.run();

    cleanupAllModels(state, app.getRenderer());

    return 0;
}
